File: src/log_migration/authoring/frontmatter.py
from datetime import date, datetime
from pathlib import Path
from typing import Any

import yaml

from .models import PageDocument, PageProblem


_REQUIRED = {"id", "page_type", "status", "created_at", "updated_at"}


def _problem(path: Path, detail: str) -> PageProblem:
    return PageProblem(path=path, detail=detail)


def parse_document(path: Path, text: str) -> PageDocument | PageProblem:
    if not text.startswith("---\n"):
        return _problem(path, "frontmatter is missing")
    marker = text.find("\n---\n", 4)
    if marker < 0:
        return _problem(path, "frontmatter is not closed")
    try:
        values = yaml.safe_load(text[4:marker])
    except yaml.YAMLError as error:
        return _problem(path, f"frontmatter is invalid: {error}")
    if not isinstance(values, dict):
        return _problem(path, "frontmatter must be a mapping")
    if "id" not in values:
        return _problem(path, "missing required key: id")
    if "page_type" in values and values["page_type"] not in ("date", "named"):
        return _problem(path, "page_type must be date or named")
    if "status" in values and values["status"] not in ("draft", "published"):
        return _problem(path, "status must be draft or published")
    missing = _REQUIRED - values.keys()
    if missing:
        return _problem(path, f"missing required key: {sorted(missing)[0]}")
    if not isinstance(values["id"], str) or not values["id"]:
        return _problem(path, "id must be a non-empty string")
    if not all(isinstance(values[key], datetime) for key in ("created_at", "updated_at")):
        return _problem(path, "created_at and updated_at must be datetimes")
    page_date = values.get("page_date")
    if isinstance(page_date, datetime) or (
        page_date is not None and not isinstance(page_date, date)
    ):
        return _problem(path, "page_date must be a date")
    published_at = values.get("published_at")
    if published_at is not None and not isinstance(published_at, datetime):
        return _problem(path, "published_at must be a datetime")
    name = values.get("name")
    title = values.get("title")
    if name is not None and not isinstance(name, str):
        return _problem(path, "name must be a string")
    if title is not None and not isinstance(title, str):
        return _problem(path, "title must be a string")
    if values["page_type"] == "date" and page_date is None:
        return _problem(path, "date pages require page_date")
    if values["page_type"] == "named" and not name:
        return _problem(path, "named pages require name")
    return PageDocument(
        id=values["id"],
        page_type=values["page_type"],
        name=name,
        page_date=page_date,
        title=title,
        status=values["status"],
        created_at=values["created_at"],
        updated_at=values["updated_at"],
        published_at=published_at,
        path=path,
        body=text[marker + len("\n---\n") :],
    )


def serialize_document(document: PageDocument) -> str:
    values: dict[str, Any] = {
        "id": document.id,
        "page_type": document.page_type,
        "name": document.name,
        "page_date": document.page_date,
        "title": document.title,
        "status": document.status,
        "created_at": document.created_at,
        "updated_at": document.updated_at,
        "published_at": document.published_at,
    }
    frontmatter = yaml.safe_dump(values, sort_keys=False, allow_unicode=True)
    return f"---\n{frontmatter}---\n{document.body}"

File: lib/weblog_authoring/frontmatter.rb
# frozen_string_literal: true

require "date"
require "pathname"
require "psych"

module WeblogAuthoring
  REQUIRED_KEYS = %w[id page_type status created_at updated_at].freeze
  ALLOWED_KEYS = (REQUIRED_KEYS + %w[name page_date title published_at]).freeze

  module_function

  def parse_document(path, text)
    path = Pathname(path)
    return problem(path, "frontmatter is missing") unless text.start_with?("---\n")

    marker = text.index("\n---\n", 4)
    return problem(path, "frontmatter is not closed") if marker.nil?

    values = load_frontmatter(path, text[4...marker])
    return values if values.is_a?(PageProblem)
    return problem(path, "frontmatter must be a mapping") unless values.is_a?(Hash)

    unknown_keys = values.keys - ALLOWED_KEYS
    return problem(path, "unknown key: #{unknown_keys.sort.first}") unless unknown_keys.empty?

    missing_keys = REQUIRED_KEYS - values.keys
    return problem(path, "missing required key: #{missing_keys.sort.first}") unless missing_keys.empty?

    return problem(path, "id must be a non-empty string") unless values["id"].is_a?(String) && !values["id"].empty?
    return problem(path, "page_type must be date or named") unless %w[date named].include?(values["page_type"])
    return problem(path, "status must be draft or published") unless %w[draft published].include?(values["status"])
    return problem(path, "created_at and updated_at must be datetimes") unless time_value?(values["created_at"]) && time_value?(values["updated_at"])

    page_date = values["page_date"]
    return problem(path, "page_date must be a date") unless page_date.nil? || page_date.instance_of?(Date)

    published_at = values["published_at"]
    return problem(path, "published_at must be a datetime") unless published_at.nil? || time_value?(published_at)

    name = values["name"]
    return problem(path, "name must be a string") unless name.nil? || name.is_a?(String)

    title = values["title"]
    return problem(path, "title must be a string") unless title.nil? || title.is_a?(String)

    if values["page_type"] == "date" && page_date.nil?
      return problem(path, "date pages require page_date")
    end
    if values["page_type"] == "named" && (name.nil? || name.empty?)
      return problem(path, "named pages require name")
    end

    PageDocument.new(
      id: values["id"],
      page_type: values["page_type"],
      name:,
      page_date:,
      title:,
      status: values["status"],
      created_at: values["created_at"],
      updated_at: values["updated_at"],
      published_at:,
      path:,
      body: text[(marker + "\n---\n".length)..] || "",
      links: []
    )
  end

  def serialize_document(document)
    values = {
      "id" => document.id,
      "page_type" => document.page_type,
      "name" => document.name,
      "page_date" => document.page_date,
      "title" => document.title,
      "status" => document.status,
      "created_at" => document.created_at&.dup,
      "updated_at" => document.updated_at&.dup,
      "published_at" => document.published_at&.dup
    }
    serialized = Psych.safe_dump(values, permitted_classes: [Date, Time], line_width: -1).delete_prefix("---\n")

    +"---\n#{serialized}---\n#{document.body}"
  end

  def load_frontmatter(path, text)
    Psych.safe_load(text, permitted_classes: [Date, Time], aliases: false)
  rescue Psych::Exception => error
    problem(path, "frontmatter is invalid: #{error.message}")
  end
  private_class_method :load_frontmatter

  def problem(path, detail)
    PageProblem.new(path:, detail:)
  end
  private_class_method :problem

  def time_value?(value)
    value.is_a?(Time)
  end
  private_class_method :time_value?
end
